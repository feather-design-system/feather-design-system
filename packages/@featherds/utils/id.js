const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
