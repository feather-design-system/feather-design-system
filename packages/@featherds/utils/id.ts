const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
