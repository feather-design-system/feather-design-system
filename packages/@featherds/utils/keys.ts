const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
