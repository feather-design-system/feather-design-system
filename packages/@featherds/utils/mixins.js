var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
