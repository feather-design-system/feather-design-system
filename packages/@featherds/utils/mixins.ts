const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
