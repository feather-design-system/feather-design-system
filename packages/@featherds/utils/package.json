{
  "name": "@featherds/utils",
  "version": "0.12.31",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {
    "compile": "tsc --declaration"
  },
  "keywords": [
    "vue",
    "FeatherUtils"
  ]
}
