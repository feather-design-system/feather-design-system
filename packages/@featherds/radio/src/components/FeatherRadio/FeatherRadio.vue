<template>
  <div class="layout-container">
    <div
      class="feather-radio"
      role="radio"
      ref="input"
      :aria-checked="checked ? 'true' : 'false'"
      :aria-disabled="disabled ? 'true' : 'false'"
      :aria-labelledby="labelId"
      :tabindex="tabindex"
      @click="click"
      @blur="blur"
      data-ref-id="feather-radio"
    >
      <div class="radio hover focus">
        <div class="icon" aria-hidden="true">
          <svg height="16" width="16" class="dot">
            <circle cx="8" cy="8" r="5" stroke="black" stroke-width="0" />
          </svg>
        </div>
        <feather-ripple center v-if="!disabled" />
      </div>

      <span class="label" data-ref-id="feather-radio-label" :id="labelId">
        <slot />
      </span>
    </div>
  </div>
</template>
<script>
import { getSafeId } from "@featherds/utils/id";
import { FeatherRipple } from "@featherds/ripple";
import { computed, inject, ref } from "vue";
export default {
  props: {
    value: undefined,
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  setup(props) {
    const first = ref(false);
    const checked = ref(false);
    const labelId = computed(() => getSafeId("radio-label-id"));
    const tabindex = computed(() => {
      if (first.value) {
        return 0;
      }
      return checked.value ? 0 : -1;
    });

    const input = ref(null);
    const focus = () => {
      input.value.focus();
    };

    //register
    const register = inject("register");
    const blur = inject("blur");
    const select = inject("select");

    const vm = {
      first,
      focus,
      disabled: props.disabled,
      value: props.value,
      checked,
    };
    register(vm);
    const click = () => {
      select(vm);
    };

    return {
      labelId,
      tabindex,
      first,
      blur,
      click,
      input,
      checked,
    };
  },
  components: {
    FeatherRipple,
  },
};
</script>

<style lang="scss" scoped>
@import "~@featherds/styles/mixins/typography";
@import "~@featherds/styles/themes/variables";
@import "~@featherds/styles/themes/utils";

.feather-radio {
  &:focus {
    outline: 0;
  }
  @include state-on-surface();
}
.layout-container {
  display: inline-block;
  line-height: 0;
}

.feather-radio {
  cursor: pointer;
  display: inline-flex;
  align-items: center;
  margin-left: -8px;
  .radio {
    height: 36px;
    width: 36px;
    display: inline-block;
    padding: 8px;
    border-radius: 100%;
  }
  .icon {
    height: 20px;
    width: 20px;
    border: 2px solid var($secondary-text-on-surface);
    background-color: transparent;
    border-radius: 100%;
    position: relative;
  }
  .dot {
    position: absolute;
    height: 16px;
    width: 16px;
    fill: var($primary);
    border-radius: 100%;
    opacity: 0;
    transform: scale(0);
  }
  .label {
    @include body-small();
    color: var($primary-text-on-surface);
    min-width: 24px;
    padding-right: 12px;
    padding-left: 4px;
  }
}

.feather-radio:hover {
  &[aria-disabled="true"] {
    cursor: default;
  }
}

.feather-radio[aria-checked="true"] {
  .icon {
    border-color: var($primary);
  }
  .dot {
    transition: all 280ms ease-in-out;
    opacity: 1;
    transform: scale(1);
  }
}
.feather-radio[aria-disabled="true"] {
  @include state-disabled();
  .icon {
    border-color: var($disabled-text-on-surface);
  }
  .label {
    color: var($disabled-text-on-surface);
  }
  .dot {
    fill: var($disabled-text-on-surface);
  }
}
</style>
