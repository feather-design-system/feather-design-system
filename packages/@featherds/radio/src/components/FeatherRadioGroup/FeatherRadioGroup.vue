<template>
  <div
    v-bind="inherittedAttrs"
    class="feather-radio-group-container"
    :class="{ vertical: vertical }"
  >
    <label :for="groupId" class="group-label">{{ label }}</label>
    <div
      v-bind="attrs"
      class="feather-radio-group"
      data-ref-id="feather-radio-group"
      role="radiogroup"
      :id="groupId"
      @keydown="keydown"
    >
      <slot />
    </div>
    <InputSubText :id="descriptionId" />
  </div>
</template>
<script>
import { getSafeId } from "@featherds/utils/id";
import {
  InputInheritAttrsMixin,
  InputSubText,
  InputSubTextMixin,
} from "@featherds/input-helper";
import { computed, toRef } from "vue";
import { useRadioGroup } from "@featherds/composables/radio/RadioGroup";
export default {
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  mixins: [InputInheritAttrsMixin, InputSubTextMixin],
  props: {
    label: {
      type: String,
      required: true,
    },
    modelValue: {
      type: [String, Number, Boolean, Array, Object, Date, Function],
    },
    vertical: {
      type: Boolean,
      default: false,
    },
  },
  setup(props, context) {
    const error = toRef(props, "error");
    const modelValue = toRef(props, "modelValue");
    const groupId = computed(() => {
      return getSafeId("feather-radio-group");
    });
    const descriptionId = computed(() => {
      return getSafeId("feather-input-description");
    });
    const attrs = computed(() => {
      const _attrs = {
        ...context.attrs,
        class: "",
        "aria-describedby": descriptionId.value,
      };
      //use aria-invalid if passed in (some validation libraries will specify this)
      //otherwise base it off this.error
      if (!_attrs["aria-invalid"]) {
        _attrs["aria-invalid"] = !!error.value;
      }
      return _attrs;
    });

    return {
      groupId,
      descriptionId,
      attrs,
      ...useRadioGroup(modelValue, context.emit),
    };
  },
  components: {
    InputSubText,
  },
};
</script>

<style lang="scss" scoped>
@import "@featherds/styles/mixins/typography";
@import "@featherds/input-helper/scss/spacing";
@import "@featherds/styles/themes/variables";

.group-label {
  @include subtitle1();
  display: block;
  color: var($primary-text-on-surface);
  min-width: 24px;
}
.feather-radio-group-container {
  @include group-spacing;
}
</style>
