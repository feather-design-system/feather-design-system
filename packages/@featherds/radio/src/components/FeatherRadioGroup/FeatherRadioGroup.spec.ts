import { mount } from "@vue/test-utils";
import FeatherRadioGroup from "./FeatherRadioGroup.vue";
import FeatherRadio from "../FeatherRadio/FeatherRadio.vue";

import axe from "@featherds/utils/test/axe";
import { getCalls } from "@featherds/utils/test/calls";
import { h, nextTick } from "vue";
import * as id from "@featherds/utils/id";
import { Code } from "@featherds/utils/keys";
import { vi, expect, describe, it } from "vitest";
vi.spyOn(id, "getSafeId").mockImplementation((x) => x);

const selector = "[role='radiogroup']";
const getWrapper = function (options: Record<string, unknown> = {}) {
  const props = (options.props as Record<string, unknown>) || {};
  options.props = {
    ...props,
    label: "Unit Test Radio Group",
  };
  return mount(FeatherRadioGroup, options);
};

const getRadio = (value: number | string, disabled: boolean) => ({
  render() {
    return h(
      FeatherRadio,
      {
        value: value,
        disabled: disabled,
      },
      {
        default: () => "Test Radio",
      }
    );
  },
});

describe("FeatherRadioGroup.vue", () => {
  it("should set id and name", () => {
    const wrapper = getWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should stack vertically", () => {
    const wrapper = getWrapper({
      props: {
        hint: "TEST HINT",
        vertical: true,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
  it("should have hint text", () => {
    const wrapper = getWrapper({
      props: {
        hint: "TEST HINT",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should have error text", async () => {
    const wrapper = getWrapper();
    await wrapper.setProps({
      error: "TEST HINT",
    });
    await nextTick();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should select non null or undefined falsey value", async () => {
    const slots = {
      default: [getRadio(0, false), getRadio(1, false)],
    };
    const wrapper = getWrapper({
      slots,
      propsData: { modelValue: 0 },
    });
    await nextTick();
    await nextTick();
    expect(
      wrapper.findAll("[role='radio']")[0].attributes("aria-checked")
    ).toBe("true");
  });

  it("should set first property on first radio if none match the value", async () => {
    const slots = {
      default: [getRadio(1, false), getRadio(2, false)],
    };
    const wrapper = getWrapper({
      slots,
      props: { modelValue: -1 },
    });
    await nextTick();
    await nextTick();
    expect(wrapper.findAll("[role='radio']")[0].attributes("tabindex")).toBe(
      "0"
    );
  });
  it("should set first property on first radio not disabled if none match the value", async () => {
    const slots = {
      default: [getRadio(1, true), getRadio(2, false)],
    };
    const wrapper = getWrapper({
      slots,
      props: { modelValue: -1 },
    });
    await nextTick();
    await nextTick();
    expect(wrapper.findAll("[role='radio']")[1].attributes("tabindex")).toBe(
      "0"
    );
  });
  it("should set select first radio that matches initial value", async () => {
    const slots = {
      default: [getRadio(1, false), getRadio("2", false)],
    };
    const wrapper = getWrapper({
      slots,
      props: { modelValue: "2" },
    });
    await nextTick();
    await nextTick();
    await nextTick();
    expect(
      wrapper.findAll("[role='radio']")[1].attributes("aria-checked")
    ).toBe("true");
  });
  it("should emit input on check", async () => {
    const slots = {
      default: [getRadio(1, false), getRadio(2, false)],
    };
    const wrapper = getWrapper({
      slots,
      props: { modelValue: 2 },
    });
    await wrapper.find("[role='radio']").trigger("click");
    expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(1);
  });

  it("should select clicked radio button", async () => {
    const slots = {
      default: [getRadio(1, false), getRadio(2, false)],
    };
    const wrapper = getWrapper({
      slots,
      props: { modelValue: -1 },
    });
    await nextTick();
    await nextTick();
    await wrapper.find("[role='radio']").trigger("click");
    expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(1);
  });

  it("should select radio button when modelValue is updated", async () => {
    const slots = {
      default: [getRadio(1, false), getRadio(2, false)],
    };
    const wrapper = getWrapper({
      slots,
      propsData: { modelValue: 1 },
    });
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    await wrapper.setProps({
      modelValue: 2,
    });
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();

    expect(
      wrapper.findAll("[role='radio']")[1].attributes("aria-checked")
    ).toBe("true");
  });
  it("should not select a disabled clicked radio button", async () => {
    const slots = {
      default: [getRadio(1, true), getRadio(2, false)],
    };
    const wrapper = getWrapper({
      slots,
      props: { modelValue: -1 },
    });
    await nextTick();
    await nextTick();
    await wrapper.find("[role='radio']").trigger("click");
    expect(wrapper.emitted("update:modelValue")).toBeUndefined();
  });
  it("should select current radio when space or enter is pressed", async () => {
    const testSelection = async (code: Code) => {
      const slots = {
        default: [getRadio(1, false), getRadio(2, false)],
      };
      const wrapper = getWrapper({
        slots,
        props: { modelValue: -1 },
      });
      await nextTick();
      await nextTick();
      const radio = wrapper.find(selector);
      await radio.trigger("keydown", { code });
      expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(1);
    };
    await testSelection(Code.ENTER);
    await testSelection(Code.SPACE);
  });
  it("should select next enabled radio when right or down is pressed", async () => {
    const testNext = async (code: Code) => {
      const slots = {
        default: [getRadio(1, false), getRadio(2, true), getRadio(3, false)],
      };
      const wrapper = getWrapper({
        slots,
        props: { modelValue: 1 },
      });
      await nextTick();
      await nextTick();
      const radio = wrapper.find(selector);
      await radio.trigger("keydown", { code });
      expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(3);
    };
    await testNext(Code.DOWN);
    await testNext(Code.RIGHT);
  });
  it("should select first enabled radio when right or down is pressed at the end of the group", async () => {
    const testNext = async (code: Code) => {
      const slots = {
        default: [getRadio(1, true), getRadio(2, false), getRadio(3, false)],
      };
      const wrapper = getWrapper({
        slots,
        props: { modelValue: 3 },
      });
      await nextTick();
      await nextTick();
      const radio = wrapper.find(selector);
      await radio.trigger("keydown", { code });
      expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(2);
    };
    await testNext(Code.DOWN);
    await testNext(Code.RIGHT);
  });
  it("should select previous enabled radio when left or up is pressed", async () => {
    const testPrev = async (code: Code) => {
      const slots = {
        default: [getRadio(1, false), getRadio(2, true), getRadio(3, false)],
      };
      const wrapper = getWrapper({
        slots,
        props: { modelValue: 3 },
      });

      await nextTick();
      await nextTick();
      const radio = wrapper.find(selector);
      await radio.trigger("keydown", { code });
      expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(1);
    };
    await testPrev(Code.LEFT);
    await testPrev(Code.UP);
  });
  it("should select last enabled radio when left or up is pressed at the start of the group", async () => {
    const testPrev = async (code: Code) => {
      const slots = {
        default: [
          getRadio(1, false),
          getRadio(0, false),
          getRadio(2, false),
          getRadio(3, true),
        ],
      };
      const wrapper = getWrapper({
        slots,
        props: { modelValue: 1 },
      });

      await nextTick();
      await nextTick();
      const radio = await wrapper.find(selector);
      await radio.trigger("keydown", { code });
      expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(2);
    };
    await testPrev(Code.LEFT);
    await testPrev(Code.UP);
  });
  describe("a11y", () => {
    it("should be accessible ", async () => {
      const slots = {
        default: [getRadio(1, false)],
      };
      const wrapper = getWrapper({
        slots,
        props: { modelValue: 1 },
      });
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
  });
});
