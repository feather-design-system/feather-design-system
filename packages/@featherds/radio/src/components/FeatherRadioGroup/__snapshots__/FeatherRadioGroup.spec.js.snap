// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatherRadioGroup.vue should have error text 1`] = `
<div
  class="feather-radio-group-container"
>
  <label
    class="group-label"
    for="feather-radio-group"
  >
    Unit Test Radio Group
  </label>
  <div
    aria-describedby="feather-input-description"
    aria-invalid="true"
    class="feather-radio-group"
    data-ref-id="feather-radio-group"
    id="feather-radio-group"
    role="radiogroup"
  >
    
    
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-input-description"
    style=""
  >
    <!---->
    <div
      aria-live="assertive"
      class="feather-input-error"
      data-ref-id="feather-form-element-error"
    >
      TEST HINT
    </div>
    
    
  </div>
</div>
`;

exports[`FeatherRadioGroup.vue should have hint text 1`] = `
<div
  class="feather-radio-group-container"
>
  <label
    class="group-label"
    for="feather-radio-group"
  >
    Unit Test Radio Group
  </label>
  <div
    aria-describedby="feather-input-description"
    aria-invalid="false"
    class="feather-radio-group"
    data-ref-id="feather-radio-group"
    id="feather-radio-group"
    role="radiogroup"
  >
    
    
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-input-description"
  >
    <div
      class="feather-input-hint"
      data-ref-id="feather-form-element-hint"
    >
      TEST HINT
    </div>
    <!---->
    
    
  </div>
</div>
`;

exports[`FeatherRadioGroup.vue should set id and name 1`] = `
<div
  class="feather-radio-group-container"
>
  <label
    class="group-label"
    for="feather-radio-group"
  >
    Unit Test Radio Group
  </label>
  <div
    aria-describedby="feather-input-description"
    aria-invalid="false"
    class="feather-radio-group"
    data-ref-id="feather-radio-group"
    id="feather-radio-group"
    role="radiogroup"
  >
    
    
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-input-description"
    style="display: none;"
  >
    <!---->
    <!---->
    
    
  </div>
</div>
`;

exports[`FeatherRadioGroup.vue should stack vertically 1`] = `
<div
  class="feather-radio-group-container vertical"
>
  <label
    class="group-label"
    for="feather-radio-group"
  >
    Unit Test Radio Group
  </label>
  <div
    aria-describedby="feather-input-description"
    aria-invalid="false"
    class="feather-radio-group"
    data-ref-id="feather-radio-group"
    id="feather-radio-group"
    role="radiogroup"
  >
    
    
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-input-description"
  >
    <div
      class="feather-input-hint"
      data-ref-id="feather-form-element-hint"
    >
      TEST HINT
    </div>
    <!---->
    
    
  </div>
</div>
`;
